File: app/src/main/java/com/arjanvlek/oxygenupdater/Download/DownloadProgressData.java
package com.arjanvlek.oxygenupdater.Download;


import android.app.Activity;

import com.arjanvlek.oxygenupdater.R;

import static com.arjanvlek.oxygenupdater.Download.UpdateDownloader.NOT_SET;

public class DownloadProgressData {

    private TimeRemaining timeRemaining;

    private int progress;


    public DownloadProgressData(long numberOfSecondsRemaining, int progress) {
        this.timeRemaining = calculateTimeRemaining((int)numberOfSecondsRemaining);
        this.progress = progress;
    }

    public TimeRemaining getTimeRemaining() {
        return timeRemaining;
    }

    public int getProgress() {
        return progress;
    }

    private TimeRemaining calculateTimeRemaining(int numberOfSecondsRemaining) {
        if(numberOfSecondsRemaining == NOT_SET) {
            return null;
        }

        return new TimeRemaining(numberOfSecondsRemaining / 3600, numberOfSecondsRemaining / 60, numberOfSecondsRemaining % 60);
    }

    public class TimeRemaining {

        private int hoursRemaining;
        private int minutesRemaining;
        private int secondsRemaining;

        public TimeRemaining(int hoursRemaining, int minutesRemaining, int secondsRemaining) {
            this.hoursRemaining = hoursRemaining;
            this.minutesRemaining = minutesRemaining;
            this.secondsRemaining = secondsRemaining;
        }

        public int getHoursRemaining() {
            return hoursRemaining;
        }

        public int getMinutesRemaining() {
            return minutesRemaining;
        }

        public int getSecondsRemaining() {
            return secondsRemaining;
        }

        public String toString(Activity activity) {
            if (getHoursRemaining() > 1) {
                return activity.getString(R.string.download_progress_text_hours_remaining, getProgress(), getHoursRemaining());
            } else if (getHoursRemaining() == 1) {
                return activity.getString(R.string.download_progress_text_one_hour_remaining, getProgress());
            } else if (getHoursRemaining() == 0 && getMinutesRemaining() > 1) {
                return activity.getString(R.string.download_progress_text_minutes_remaining, getProgress(), getMinutesRemaining());
            } else if (getHoursRemaining() == 0 && getMinutesRemaining() == 1) {
                return activity.getString(R.string.download_progress_text_one_minute_remaining, getProgress());
            } else if (getHoursRemaining() == 0 && getMinutesRemaining() == 0 && getSecondsRemaining() > 10) {
                return activity.getString(R.string.download_progress_text_less_than_a_minute_remaining, getProgress());
            } else if (getHoursRemaining() == 0 && getMinutesRemaining() == 0 && getSecondsRemaining() <= 10) {
                return activity.getString(R.string.download_progress_text_seconds_remaining, getProgress());
            } else {
                return activity.getString(R.string.download_progress_text_unknown_time_remaining, getProgress());
            }
        }
    }
}

File: app/src/main/java/com/arjanvlek/oxygenupdater/models/DownloadProgressData.kt
package com.arjanvlek.oxygenupdater.models

import android.content.Context
import com.arjanvlek.oxygenupdater.R
import com.arjanvlek.oxygenupdater.services.DownloadService
import java.io.Serializable

/**
 * Contains the progress & ETA of a download.
 */
data class DownloadProgressData(val numberOfSecondsRemaining: Long, val progress: Int, val isWaitingForConnection: Boolean = false) : Serializable {

    val timeRemaining = calculateTimeRemaining(numberOfSecondsRemaining.toInt())

    constructor(numberOfSecondsRemaining: Long, progress: Int) : this(numberOfSecondsRemaining, progress, false)

    private fun calculateTimeRemaining(numberOfSecondsRemaining: Int): TimeRemaining? {
        return if (numberOfSecondsRemaining == DownloadService.NOT_SET.toInt()) {
            null
        } else {
            TimeRemaining(numberOfSecondsRemaining / 3600, numberOfSecondsRemaining / 60, numberOfSecondsRemaining % 60)
        }
    }

    /**
     * Can't be private, because UpdateInformationFragment calls this.
     */
    inner class TimeRemaining internal constructor(val hoursRemaining: Int, val minutesRemaining: Int, val secondsRemaining: Int) : Serializable {

        fun toString(context: Context?): String {
            if (context == null) {
                return ""
            }

            return if (hoursRemaining > 1) {
                context.getString(R.string.download_progress_text_hours_remaining, hoursRemaining)
            } else if (hoursRemaining == 1) {
                context.getString(R.string.download_progress_text_one_hour_remaining)
            } else if (hoursRemaining == 0 && minutesRemaining > 1) {
                context.getString(R.string.download_progress_text_minutes_remaining, minutesRemaining)
            } else if (hoursRemaining == 0 && minutesRemaining == 1) {
                context.getString(R.string.download_progress_text_one_minute_remaining)
            } else if (hoursRemaining == 0 && minutesRemaining == 0 && secondsRemaining > 10) {
                context.getString(R.string.download_progress_text_less_than_a_minute_remaining)
            } else if (hoursRemaining == 0 && minutesRemaining == 0 && secondsRemaining <= 10) {
                context.getString(R.string.download_progress_text_seconds_remaining)
            } else {
                context.getString(R.string.download_progress_text_unknown_time_remaining)
            }
        }
    }

    companion object {
        private const val serialVersionUID = 5271414164650145215L
    }
}
